Feat: Global config file

logseq-cldwalker opened the pull request:
This PR introduces a global config file, ~/.logseq/config/config.edn, for desktop apps as described in [this task](https://quire.io/w/Logseq-_team_dedicated_to_sport_of_knowledge/815/Global_config_file_for_desktop?filter=all). The global config file is editable in Settings, like the graph-specific config file:
<img width="668" alt="Screen Shot 2022-09-05 at 5 25 08 PM" src="https://user-images.githubusercontent.com/97210743/188513643-4bd0c785-a0e0-4ae0-9027-7158e3752d62.png">

Behaviors that can be confirmed with QA:
- The global config file (and directory) is created from a default template if a user does not have one
- Global config files is loaded correctly on startup
- Global config state is updated in the app when a user edits the global config file outside the app
- When user switches to a new graph, global config state is applied to the current graph (db)
- When window reloads, global file watcher still works
- When multiple windows are open, global file watcher works for each window

Global config file is overridden by a graph config file except for maps which are merged together. See [the template file](https://github.com/logseq/logseq/pull/6531/files#diff-352be2ff560b02bc3bedc4d7cdfb51c3837872e186dbf95ded1c838ca390fe0a) for more explanation and tests for specific examples.

This PR has the following refactoring:
- Organized `frontend.state` so it is clear what fns are static config and which ones are reactive for use with the UI
- Introduced system components with `frontend.handler.{graph-config,repo-config}`. This is with the goal of getting logseq to a manageable system [as described in this task](https://quire.io/w/Logseq-_team_dedicated_to_sport_of_knowledge/1378/Manage_app_components_with_a_system?filter=all)
- All config in the app is read through `state/{get-config,sub-config}` so that global config can work correctly

As for the location of the config file, `~/.logseq/config/config.edn`, I debated to use it or `~/.logseq/config.edn`. Here are the pros and cons I saw with this decision. Interested to hear what others think of config location.
- Pros:
   - There is a dedicated directory for user config. This makes it easy to add new config files 
   - `~/.logseq/config` is less confusing for a user to look in as `~/.logseq` is filled with internal (not user facing) directories
   - The file watcher only watches the user config directory. If we watch `~/.logseq` we will need to filter out all file events from that directory. I don't know if there would be any weird behavior with plugins or persisting graphs to disk which all occur under `~/.logseq`
- Cons:
   - `~/.logseq/config/config.edn` is a less common config path then `~/.logseq/config.edn`

logseq-cldwalker commented on src/main/frontend/state.cljs:
```diff
@@ -520,6 +431,242 @@
+;; ============================================
+
+;; State that some subs are dependent on. Should they use sub too?
+(declare get-edit-input-id document-mode?)
```
It was surprising to see two reactive fns that depended on these non-reactive fns. Should these two have reactive versions?

tiensonqin replied:
Thanks for pointing out those invalid usage!

For `get-edit-input-id`, it's called in `sub-edit-content` which is only used for the `mock-textarea`, I changed it to subscribe to the passed input `id` instead of using `get-edit-input-id` in 2f8acf439.

For `document-mode`, it's used in `doc-mode-enter-for-new-line?` which is called when there's any keydown or keyup event, so there's no need to be reactive. I fixed it in 133902744.

logseq-cldwalker replied:
Good to know 👍

tiensonqin reviewed (approved):
So exciting to see the global config support! 🥳  Love all the refactoring too! 👍 

I QAed this PR and found there's one problem that shortcuts defined in the global config don't work when starting logseq. After some debugging, I think the reason is that this PR relies on the file watcher to get the content of the global config, which could be after the call `shortcut/refresh!`. I changed it to read the config file through `fs` to make sure it happens before the `shortcut/refresh!`. 

I also made some changes to refresh shortcuts after updating the config in Logseq.

andelf commented on src/main/frontend/fs/protocol.cljs:
```diff
@@ -15,7 +15,7 @@
   (open-dir [this ok-handler])
   (get-files [this path-or-handle ok-handler])
-  (watch-dir! [this dir])
+  (watch-dir! [this dir options])
```
This is an electron-only option, better eliminate it.

logseq-cldwalker replied:
This is needed for https://github.com/logseq/logseq/pull/6531/files#diff-66ac3108ca57e1e347706a8e0631a62cb312c54132bedd09401f9197883736bbL174 which is used by https://github.com/logseq/logseq/pull/6531/files#r963700179

andelf replied:
The protocol should hide the differences across all platforms instead of introducing a platform-specific option.
I guess the Global config is not bonded to the current graph(or else it should be saved to the current graph directory). So the change event should be broadcast to all windows.

logseq-cldwalker replied:
> The protocol should hide the differences across all platforms instead of introducing a platform-specific option

This feature can also be supported on mobile eventually with a config repo like https://github.com/cldwalker/logseq-config. Regardless, the option has been changed to be feature agnostic and can be used by any global directory

> The protocol should hide the differences across all platforms instead of introducing a platform-specific option.
I guess the Global config is not bonded to the current graph(or else it should be saved to the current graph directory). So the change event should be broadcast to all windows.

Good idea on broadcasting to all windows! I had forgotten to QA multi window cases and found it was broken. This is now fixed

andelf commented on src/main/frontend/handler.cljs:
```diff
@@ -79,45 +81,48 @@
+           (ui-handler/add-style-if-exists!)
+
+           (->
+            (p/let [_ (repo-config-handler/start {:repo repo})
+                    _ (when (config/global-config-enabled?)
+                        (global-config-handler/start {:repo repo}))])
```
```suggestion
            (p/do! (repo-config-handler/start {:repo repo})
                   (when (config/global-config-enabled?)
                        (global-config-handler/start {:repo repo})))
```

logseq-cldwalker commented on src/main/frontend/handler/global_config.cljs:
```diff
@@ -0,0 +1,80 @@
+(defn- watch-dir!
+  "Watches global config dir for given repo/db"
+  [repo]
+  (fs/watch-dir! (global-config-dir)
```
permalink

logseq-cldwalker commented:
> I QAed this PR and found there's one problem that shortcuts defined in the global config don't work when starting logseq. After some debugging, I think the reason is that this PR relies on the file watcher to get the content of the global config, which could be after the call `shortcut/refresh!`. I changed it to read the config file through `fs` to make sure it happens before the `shortcut/refresh!`.

@tiensonqin I'd seen an intermittent issue with shortcuts but saw it inconsistently enough that I started to doubt it was an issue. Thanks for fixing! Since you removed the unused :db-schema storage, I removed setting it in https://github.com/logseq/logseq/pull/6531/commits/5303aec2c449698d971b61d2dd510470a6da5b70

andelf commented on src/main/frontend/handler.cljs:
```diff
@@ -73,51 +73,48 @@
-                                 :db/version db-schema/version)))
-
 (defn restore-and-setup!
-  [repos old-db-schema]
```
Better not to change the original logic here. This breaks mobile loading.

Some fixes are merged recently to fix mobile loading and restoring.
#6561 https://github.com/logseq/logseq/commit/63d268f5f00df87b92b39e1bebc8b6274a7d0492

andelf commented on src/main/frontend/handler.cljs:
```diff
@@ -73,51 +73,48 @@
+                  (repo-handler/setup-local-repo-if-not-exists!)
+
+                  :else
+                  (state/set-db-restoring! false)))))))
```
I didn't look into the detailed cause. It seems iOS App will never clear db-restoring state now and get stuck at the splash screen.

Should test the following cases for mobile loading:

- Clear all App data, start from a fresh install
- Upgrade, all App data exists
- Changing App, all App data exits, but mobile documents path changed. This occurs when developing or switching install source from test flight to app-store.

andelf reviewed (approved):
LGTM

andelf commented on src/main/frontend/handler/global_config.cljs:
```diff
@@ -0,0 +1,65 @@
+            [cljs.reader :as reader]
+            [promesa.core :as p]
+            [shadow.resource :as rc]
+            [electron.ipc :as ipc]
```
It should not depend on electron here. This breaks mobile.
